File: java/lld/state-machines/coffee-machine/Inventory.java
import java.util.*;
import java.util.stream.Collectors;

// The resource guard. canMake() gates transitions; consume() commits.
public class Inventory {
    private final Map<String, Integer> stock;
    public Inventory(Map<String, Integer> stock) { this.stock = new HashMap<>(stock); }

    public boolean canMake(Recipe r) {
        return r.needs().entrySet().stream()
            .allMatch(e -> stock.getOrDefault(e.getKey(), 0) >= e.getValue());
    }

    public List<String> missing(Recipe r) {
        return r.needs().entrySet().stream()
            .filter(e -> stock.getOrDefault(e.getKey(), 0) < e.getValue())
            .map(Map.Entry::getKey).collect(Collectors.toList());
    }

    public void consume(Recipe r) {
        r.needs().forEach((ing, qty) -> stock.merge(ing, -qty, Integer::sum));
    }

    public void refill(String ing, int qty) { stock.merge(ing, qty, Integer::sum); }
}

File: python/lld/state-machines/coffee-machine/inventory.py
from recipe import Recipe


class Inventory:
    """The resource guard. can_make() gates transitions; consume() commits."""

    def __init__(self, stock: dict):
        self._stock = dict(stock)

    def can_make(self, r: Recipe) -> bool:
        return all(self._stock.get(ing, 0) >= qty for ing, qty in r.needs.items())

    def missing(self, r: Recipe) -> list:
        return [ing for ing, qty in r.needs.items() if self._stock.get(ing, 0) < qty]

    def consume(self, r: Recipe):
        for ing, qty in r.needs.items():
            self._stock[ing] -= qty

    def refill(self, ing: str, qty: int):
        self._stock[ing] = self._stock.get(ing, 0) + qty